src/api/server.ts

```diff
@@ -0,0 +1,114 @@
+type ServerState = StartingState | ReadyState;
+
+// A null value means that no one has requested a server yet.
+let serverProcess: ServerState | null = null;
+
+async function startServer() {
+  console.log('Starting RefStudio server');
+  const command = new TauriCommand('call-sidecar', ['serve']);
+  command.stderr.addListener('data', (line) => {
+    console.log('server stderr', line);
+  });
+  command.stdout.addListener('data', (line) => {
+    console.log('server stdout', line);
+  });
+  command.addListener('close', (data: Pick<ChildProcess, 'code' | 'signal'>) => {
+    console.warn('server closed, restarting', data.code, data.signal);
+    serverProcess = null;
+    void getOrStartServer();
+  });
+  command.addListener('error', (data) => {
+    console.error('server crashed, restarting', data);
+    serverProcess = null;
+    void getOrStartServer();
+  });
+  return command.spawn();
+}
+
+/** Hit the status endpoint on the server to determine whether it's running and healthy. */
+async function isServerHealthy() {
+  try {
+    const response = await tauriFetch(`${REFSTUDIO_HOST}/api/meta/status`);
+    if (!response.ok) {
+      return false;
+    }
+    console.log(response.data);
+    const { status } = response.data as { status: string };
+    if (status !== 'ok') {
+      return false;
+    }
+    return true;
+  } catch (e) {
+    return false;
+  }
+}
+
+/** Kill an existing server under our management if there is one. This will trigger a restart. */
+async function killServer() {
+  if (serverProcess?.state === 'ready') {
+    await serverProcess.process.kill();
+  }
+}
+
+if (import.meta.env.DEV) {
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  (window as any).killServer = killServer;
+}
+
+/** Get a reference to an existing server under our management, starting one if necessary. */
+async function getOrStartServer() {
+  if (serverProcess) {
+    if (serverProcess.state === 'ready') {
+      return serverProcess.process;
+    }
+    return null;
+  }
+  serverProcess = { state: 'starting' };
+
+  // Check if there's already a server process running. If so, kill it and start a new one.
+  if (await isServerHealthy()) {
+    console.log('Found existing API server; killing it before starting a new one.');
+    try {
+      await tauriFetch(`${REFSTUDIO_HOST}/api/meta/shutdown`, { method: 'POST' });
+    } catch (e) {
+      // Failure is expected here since the server shuts down before responding.
+    }
+  }
+
+  const process = await startServer();
+  serverProcess = { state: 'ready', process };
+  return process;
+}
+
+/**
+ * Ensure there is exactly one RefStudio server running.
+ *
+ * Repeated calls have no effect. The server will terminate when the RefStudio app terminates.
+ * If an existing server not managed by us is already running, it will be killed. (This can happen
+ * when you run Tauri Desktop in dev mode and make TypeScript changes.)
+ */
+export function runRefStudioServerOnDesktop() {
```

Comment:
There's a possible race condition where we start the server and try to fetch something from it (maybe settings) before it's ready to respond. I'm hoping this won't come up in practice. If it is, we'll want to expose something like `function waitForServer(): Promise<void>` and `await` that before running `tauriFetch`. But this will introduce another difference between desktop and web (where you can assume the server is always running) so I'd really like to avoid that if we can.

Reply:
I think it's ok. AppStartup (and the splashscreen) is there to ensure we don't call any other API (or render the app) before everything is initialised.

Reply:
Yeah, I think we're going to need some kind of waiting logic. Here's what I see in the debug logs. Note that we make at least 3 API calls (init settings, ingest references, read file tree) before the HTTP server reports that it's listening:
<img width="1335" alt="image" src="https://github.com/refstudio/refstudio/assets/98301/2144f19d-8714-495c-913d-a7b5eb857f19">